SignalServiceKit/src/Util/Cryptography.m

```diff
@@ -386,137 +385,166 @@ + (NSData *)encryptAttachmentData:(NSData *)attachmentData
+    NSUInteger cipherTextLength = encryptedData.length - kAESGCM256_IVLength - kAESGCM256_TagLength;
+
+    // encryptedData layout: initializationVector || ciphertext || authTag
+    NSData *initializationVector = [encryptedData subdataWithRange:NSMakeRange(0, kAESGCM256_IVLength)];
+    NSData *ciphertext = [encryptedData subdataWithRange:NSMakeRange(kAESGCM256_IVLength, cipherTextLength)];
     NSData *authTag =
-        [encryptedData subdataWithRange:NSMakeRange(kAESGCM128_IVLength + cipherTextLength, kAESGCM128_TagLength)];
+        [encryptedData subdataWithRange:NSMakeRange(kAESGCM256_IVLength + cipherTextLength, kAESGCM256_TagLength)];
 
     return
-        [self decryptAESGCMWithInitializationVector:initializationVector cipherText:cipherText authTag:authTag key:key];
+        [self decryptAESGCMWithInitializationVector:initializationVector ciphertext:ciphertext authTag:authTag key:key];
 }
 
 + (nullable NSData *)decryptAESGCMWithInitializationVector:(NSData *)initializationVector
-                                                cipherText:(NSData *)cipherText
+                                                ciphertext:(NSData *)ciphertext
                                                    authTag:(NSData *)authTagFromEncrypt
-                                                       key:(OWSAES128Key *)key
+                                                       key:(OWSAES256Key *)key
 {
-    void *plainTextBytes = malloc(cipherText.length);
-    if (plainTextBytes == NULL) {
-        OWSFail(@"Failed to malloc plainTextBytes");
+    NSMutableData *plaintext = [NSMutableData dataWithLength:ciphertext.length];
+
+    // Create and initialise the context
+    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+
+    if (!ctx) {
+        OWSFail(@"%@ failed to build context while decrypting", self.tag);
+        return nil;
+    }
+
+    // Initialise the decryption operation.
+    if (EVP_DecryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL) != kOpenSSLSuccess) {
+        OWSFail(@"%@ failed to init decryption", self.tag);
+        return nil;
+    }
+
+    // Set IV length. Not necessary if this is 12 bytes (96 bits)
+    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, kAESGCM256_IVLength, NULL) != kOpenSSLSuccess) {
+        OWSFail(@"%@ failed to set key and iv while decrypting", self.tag);
+        return nil;
+    }
+
+    // Initialise key and IV
+    if (EVP_DecryptInit_ex(ctx, NULL, NULL, key.keyData.bytes, initializationVector.bytes) != kOpenSSLSuccess) {
+        OWSFail(@"%@ failed to init decryption", self.tag);
         return nil;
     }
 
-    void *decryptAuthTagBytes = malloc(kAESGCM128_TagLength);
-    if (decryptAuthTagBytes == NULL) {
-        OWSFail(@"Failed to malloc decryptAuthTagBytes");
-        free(plainTextBytes);
+    // Provide the message to be decrypted, and obtain the plaintext output.
+    // EVP_DecryptUpdate can be called multiple times if necessary
```

Comment:
What does this mean: `EVP_DecryptUpdate can be called multiple times if necessary`?  I don't see us calling it multiple times?